ground/gcs/src/plugins/config/configautotunewidget.cpp

```diff
@@ -407,23 +395,309 @@ void ConfigAutotuneWidget::stuffShareForm(AutotuneFinalPage *autotuneShareForm)
     }
 }
 
+// XXX appropriately encapsulate all this crap
+#define ATFLASH_MAGIC 0x656e755480008041 /* A...Tune */
+struct at_flash_header
+{
+    uint64_t magic;
+    uint16_t wiggle_points;
+    uint16_t aux_data_len;
+    uint16_t sample_rate;
+
+    // Consider total number of averages here
+    uint16_t resv;
+};
+
+struct at_measurement
+{
+    float y[3]; /* Gyro measurements */
+    float u[3]; /* Actuator desired */
+};
+
+struct at_flash
+{
+    struct at_flash_header hdr;
+
+    struct at_measurement data[];
+};
+
+/* Run a Butterworth biquad filter on a circular buffer.  First go around once
+ * to "prime" the filter, then actually filter in place.  This is derived from
+ * @glowtape's excellent flight implementation
+ */
+void biquad_filter_in_place(float cutoff, int pts, float *data)
+{
+    float f = 1.0f / tan(M_PI * cutoff);
+    float q = 1.4142f;
+
+    float y2 = 0, y1 = 0, x2 = 0, x1 = 0;
+
+    float b0 = 1.0f / (1.0f + q * f + f * f);
+    float a1 = 2.0f * (f * f - 1.0f) * b0;
+    float a2 = -(1.0f - q * f + f * f) * b0;
+
+    for (int i = 0; i < pts; i++) {
+        float y = b0 * (data[i] + 2.0f * x1 + x2) + a1 * y1 + a2 * y2;
+
+        y2 = y1;
+        y1 = y;
+
+        x2 = x1;
+        x1 = data[i];
+    }
+
+    for (int i = 0; i < pts; i++) {
+        float y = b0 * (data[i] + 2.0f * x1 + x2) + a1 * y1 + a2 * y2;
+
+        y2 = y1;
+        y1 = y;
+
+        x2 = x1;
+        x1 = data[i];
+
+        data[i] = y;
+    }
+}
+
+/* Returns number of samples of delay between series */
+float get_sample_delay(int pts, const QVector<float> &delayed, const QVector<float> &orig)
+{
+    ffft::FFTReal<float> fft(pts);
+
+    /* XXX avoid all these VLAs on stack */
+    /* Convert to frequency domain */
+    QVector<float> delayed_fft(pts);
```

Comment:
Keeping an eye on the Qt dev mailing list... I'm weary of using Qt containers that are more or less identical to STL containers unless they do something the STL one cannot.